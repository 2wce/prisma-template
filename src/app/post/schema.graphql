input CreateDraftInput {
  content: String
  title: String!
}

input UpdatePostInput {
  id: ID!
  content: String
  title: String
}

type Post {
  id: ID!
  title: String!
  content: String
  published: Boolean!
  author: User
}

type Query {
  posts(searchTerm: String): [Post!]!
  post(id: ID!): Post
}

type Mutation {
  createDraft(input: CreateDraftInput!): Post!
  updatePost(input: UpdatePostInput!): Post!
  deletePost(id: ID!): Post
  publish(id: ID!): Post
}
