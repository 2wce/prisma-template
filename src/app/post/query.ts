import type { QueryPostArgs, QueryPostsArgs } from "@/generated";
import type { Context } from "@/utils";
import type { Post } from "@prisma/client";

/*
 * Export an object that includes the 'post' and 'posts' modules.
 * This object can be used to map the 'post' and 'posts' queries in your GraphQL schema to their corresponding resolvers.
 */
export default {
  posts: (_parent: unknown, args: QueryPostsArgs, ctx: Context) => {
    /*
     * Define the 'where' object, which includes the conditions for the posts to get.
     * If a search term is provided, the posts to get should have a title or content that contains the search term and be published.
     * If no search term is provided, the posts to get should just be published.
     */
    const where = args.searchTerm
      ? {
          OR: [
            { title: { contains: args.searchTerm } },
            { content: { contains: args.searchTerm } },
          ],
          published: true,
        }
      : { published: true };

    /*
     * Return a promise that resolves to the posts that match the conditions in the 'where' object.
     * The 'prisma' client from the context is used to find the posts in the database.
     */
    return ctx.prisma.post.findMany({
      where,
    });
  },
  post: async (
    _parent: unknown,
    { id }: QueryPostArgs,
    { prisma }: Context,
  ): Promise<Post | null> => {
    try {
      /*
       * Try to find the post with the given ID in the database using the 'prisma' client.
       * If the post is found, return it.
       */
      return await prisma.post.findUnique({
        where: { id },
      });
    } catch (error) {
      /*
       * If an error occurs while trying to find the post, log the error and return it.
       */
      console.log(error);
      throw error as Error;
    }
  },
};
