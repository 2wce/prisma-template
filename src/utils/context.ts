import { PrismaClient } from '@prisma/client'

const prisma = new PrismaClient()
export interface Context {
  prisma: PrismaClient
  request: any
}

export function createContext(request: any) {
  return {
    ...request,
    prisma,
  }
}
