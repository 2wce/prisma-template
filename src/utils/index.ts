export * from './jwt'
export * from './prisma'
