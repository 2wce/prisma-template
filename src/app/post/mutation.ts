import type {
	MutationCreateDraftArgs,
	MutationDeletePostArgs,
	MutationPublishArgs,
	MutationUpdatePostArgs,
} from "@/generated";
import { formatError, type Context } from "@/utils";

export default {
	updatePost: async (
		_parent: unknown,
		{ input }: MutationUpdatePostArgs,
		{ prisma, userId }: Context,
	) => {
		try {
			/*
			 * Destructure the 'id', 'title', and 'content' from the 'input'.
			 */
			const { title, content, id } = input;

			/*
			 * If the ID of the post or the ID of the current user is not provided, throw an error.
			 */
			if (!id || !userId) {
				throw new Error("Invalid input");
			}

			/*
			 * Check if a post with the given ID that is linked to the current user exists in the database using the 'prisma' client.
			 * If the post does not exist, throw an error.
			 */
			const postExists = await prisma.post.findFirst({
				where: {
					AND: [{ id }, { author: { id: userId } }],
				},
			});

			if (!postExists) {
				throw new Error("Cannot update post");
			}

			/*
			 * If the post exists, return a promise that resolves to the updated post.
			 * The 'prisma' client is used to update the post in the database with the given ID and the provided title and content.
			 */
			return await prisma.post.update({
				where: { id },
				data: {
					...(title && { title }),
					...(content && { content }),
				},
			});
		} catch (error) {
			/*
			 * If an error occurs while trying to update the post, format the error and return it.
			 */
			formatError("updatePost", error as Error);
			return error;
		}
	},
	createDraft: (
		_parent: unknown,
		{ input }: MutationCreateDraftArgs,
		{ prisma, userId }: Context,
	) => {
		/*
		 * Destructure the 'title' and 'content' from the 'input'.
		 */
		const { title, content } = input;

		/*
		 * Return a promise that resolves to the created post.
		 * The 'prisma' client is used to create the post in the database with the given title and content, and it's not published yet.
		 * The author of the post is the current user, which is connected by their ID.
		 */
		return prisma.post.create({
			data: {
				title,
				content,
				published: false,
				author: {
					connect: { id: userId },
				},
			},
		});
	},
	deletePost: (
		_parent: unknown,
		{ id }: MutationDeletePostArgs,
		{ prisma }: Context,
	) => {
		/*
		 * Return a promise that resolves to the deleted post.
		 * The 'prisma' client is used to delete the post in the database with the given ID.
		 */
		return prisma.post.delete({
			where: { id },
		});
	},
	publish: (_parent: unknown, { id }: MutationPublishArgs, ctx: Context) => {
		/*
		 * Return a promise that resolves to the published post.
		 * The 'prisma' client from the context is used to update the post in the database with the given ID and set its 'published' field to true.
		 */
		return ctx.prisma.post.update({
			where: { id },
			data: { published: true },
		});
	},
};
