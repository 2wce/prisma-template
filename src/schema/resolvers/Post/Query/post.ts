import { QueryPostArgs } from '../../../../generated'
import { Context } from '../../../../utils'

export default (
  _parent: unknown,
  { id }: QueryPostArgs,
  { prisma }: Context,
) => {
  try {
    return prisma.post.findUnique({
      where: { id },
    })
  } catch (error) {
    console.log(error)
    return error
  }
}
