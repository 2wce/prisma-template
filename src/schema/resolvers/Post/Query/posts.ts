import { QueryPostsArgs } from "../../../../generated";
import { Context } from "../../../../utils";

export default (_parent: unknown, args: QueryPostsArgs, ctx: Context) => {
	const where = args.searchTerm
		? {
				OR: [
					{ title: { contains: args.searchTerm } },
					{ content: { contains: args.searchTerm } },
				],
				published: true,
		  }
		: { published: true };

	return ctx.prisma.post.findMany({
		where,
	});
};
