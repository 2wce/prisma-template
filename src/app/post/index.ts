/*
 * Import the 'Post' type from the '@prisma/client' package.
 * This type represents a post in your application.
 */
import type { Post } from "@prisma/client";

/*
 * Import the 'Context' type.
 * This type represents the context of a GraphQL resolver function, which includes any data that every resolver
 * function should have access to, like the current user or database access.
 */
import type { Context } from "@/utils";

/*
 * Import the 'Mutation' and 'Query' objects.
 * These objects include resolver functions for the mutations and queries defined in your GraphQL schema.
 */
import Mutation from "./mutation";
import Query from "./query";

/*
 * Export an object that includes the 'Query' and 'Mutation' objects and a 'Post' object.
 * The 'Post' object includes an 'author' method, which is a resolver function for getting the author of a post.
 */
export default {
  Query,
  Mutation,
  Post: {
    author: (parent: Post, _args: unknown, { prisma }: Context) => {
      /*
       * The 'author' method takes three arguments: 'parent', '_args', and 'ctx'.
       * 'parent' is the post for which to get the author.
       * '_args' includes any arguments passed to the 'author' query, but it's not used in this function, so it's named '_args'.
       * 'ctx' is the context of the resolver function, which includes the 'prisma' client.
       *
       * The function returns a promise that resolves to the author of the post.
       * It uses the 'prisma' client to find the post in the database and get its author.
       */
      return prisma.post
        .findUnique({
          where: { id: parent.id },
        })
        .author();
    },
  },
};
