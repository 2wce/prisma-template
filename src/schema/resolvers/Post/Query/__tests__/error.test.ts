import { Context, createMockContext, MockContext } from '../../../../../utils'
import post from '../post'
import posts from '../posts'

let mockCtx: MockContext
let context: Context

beforeEach(() => {
  mockCtx = createMockContext()
  context = mockCtx as unknown as Context
})

describe('Successful post queries', () => {
  it('can get filtered posts with valid search string', async () => {
    // setup
    const args = { searchTerm: 'Subscribe to' }

    // test
    const filteredPosts = await posts({}, args, context)

    // assert
    expect(filteredPosts).toMatchSnapshot()
  })

  it('can get feed with published posts', async () => {
    // test
    const publishedPosts = await posts({}, {}, context)

    // assert
    expect(publishedPosts).toMatchSnapshot()
  })

  it('can get post with valid id', async () => {
    // setup
    const args = { id: 2 }

    // test
    const result = await post({}, args, context)

    // assert
    expect(result).toMatchSnapshot()
  })
})
