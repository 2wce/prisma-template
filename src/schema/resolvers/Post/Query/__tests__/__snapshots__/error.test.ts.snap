// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Successful post queries can get feed with published posts 1`] = `undefined`;

exports[`Successful post queries can get filtered posts with valid search string 1`] = `undefined`;

exports[`Successful post queries can get post with valid id 1`] = `undefined`;
